File: back/central-reserve/internal/infra/primary/http2/handlers/reservehandler/mapper/reserve-mapper.go
package mapper

import (
	"central_reserve/internal/domain/dtos"
	"central_reserve/internal/domain/entities"
	"central_reserve/internal/infra/primary/http2/handlers/reservehandler/request"
	"central_reserve/internal/infra/primary/http2/handlers/reservehandler/response"
)

// ReserveToDomain convierte un request.Reservation a entities.Reservation
func ReserveToDomain(r request.Reservation) entities.Reservation {
	return entities.Reservation{
		BusinessID:     r.BusinessID,
		StartAt:        r.StartAt,
		EndAt:          r.EndAt,
		NumberOfGuests: r.NumberOfGuests,
	}
}

// MapToReserveDetail convierte un dtos.ReserveDetailDTO a response.ReserveDetail
func MapToReserveDetail(dto dtos.ReserveDetailDTO) response.ReserveDetail {
	return response.ReserveDetail{
		ReservaID:          dto.ReservaID,
		StartAt:            dto.StartAt,
		EndAt:              dto.EndAt,
		NumberOfGuests:     dto.NumberOfGuests,
		ReservaCreada:      dto.ReservaCreada,
		ReservaActualizada: dto.ReservaActualizada,
		EstadoCodigo:       dto.EstadoCodigo,
		EstadoNombre:       dto.EstadoNombre,
		ClienteID:          dto.ClienteID,
		ClienteNombre:      dto.ClienteNombre,
		ClienteEmail:       dto.ClienteEmail,
		ClienteTelefono:    dto.ClienteTelefono,
		ClienteDni: func() string {
			if dto.ClienteDni != nil {
				return *dto.ClienteDni
			}
			return ""
		}(),
		MesaID:           dto.MesaID,
		MesaNumero:       dto.MesaNumero,
		MesaCapacidad:    dto.MesaCapacidad,
		NegocioID:        dto.NegocioID,
		NegocioNombre:    dto.NegocioNombre,
		NegocioCodigo:    dto.NegocioCodigo,
		NegocioDireccion: dto.NegocioDireccion,
		UsuarioID:        dto.UsuarioID,
		UsuarioNombre:    dto.UsuarioNombre,
		UsuarioEmail:     dto.UsuarioEmail,
	}
}

// MapToReserveDetailList convierte un slice de dtos.ReserveDetailDTO a slice de response.ReserveDetail
func MapToReserveDetailList(dtoList []dtos.ReserveDetailDTO) []response.ReserveDetail {
	if dtoList == nil {
		return nil
	}

	responseList := make([]response.ReserveDetail, len(dtoList))
	for i, dto := range dtoList {
		responseList[i] = MapToReserveDetail(dto)
	}
	return responseList
}

File: mobile/rupu/lib/domain/infrastructure/mappers/reserve_mapper.dart
import 'package:rupu/domain/entities/reserve.dart';
import 'package:rupu/domain/infrastructure/models/reserve_response_model.dart';

class ReservesMapper {
  static ReservesResponse responseToEntity(ReservasResponseModel model) =>
      ReservesResponse(
        // <-- Asegúrate que sea *ReservasResponse*, no *ReservesResponse*
        success: model.success,
        data: model.data.map(reservaFromModel).toList(),
        filters: _filtersFromModel(model.filters),
        message: model.message,
        total: model.total,
      );

  static Reserve reservaFromModel(ReservaModel m) => Reserve(
    reservaId: m.reservaId,
    startAt: m.startAt,
    endAt: m.endAt,
    numberOfGuests: m.numberOfGuests,
    reservaCreada: m.reservaCreada,
    reservaActualizada: m.reservaActualizada,
    estadoCodigo: m.estadoCodigo,
    estadoNombre: m.estadoNombre,
    clienteId: m.clienteId,
    clienteNombre: m.clienteNombre,
    clienteEmail: m.clienteEmail,
    clienteTelefono: m.clienteTelefono,
    clienteDni: m.clienteDni,
    mesaId: m.mesaId,
    mesaNumero: m.mesaNumero,
    mesaCapacidad: m.mesaCapacidad,
    negocioId: m.negocioId,
    negocioNombre: m.negocioNombre,
    negocioCodigo: m.negocioCodigo,
    negocioDireccion: m.negocioDireccion,
    usuarioId: m.usuarioId,
    usuarioNombre: m.usuarioNombre,
    usuarioEmail: m.usuarioEmail,
    statusHistory: m.statusHistory.map(_statusHistoryFromModel).toList(),
  );

  static StatusHistoryItem _statusHistoryFromModel(StatusHistoryItemModel m) =>
      StatusHistoryItem(
        id: m.id,
        statusId: m.statusId,
        statusCode: m.statusCode,
        statusName: m.statusName,
        changedAt: m.changedAt,
        changedByUserId: m.changedByUserId,
        changedByUser: m.changedByUser,
      );

  static Filters _filtersFromModel(FiltersModel m) => Filters(
    clientId: m.clientId,
    startDate: m.startDate,
    endDate: m.endDate,
    statusId: m.statusId,
    tableId: m.tableId,
  );
}
